lib/mwUtil.js

```diff
@@ -386,4 +386,54 @@ mwUtil.isSelfRedirect = (req, location) => {
     return true;
 };
 
+/**
+ * Traverses through a request body and replaces the given keys with
+ * the content fetched using the 'fetch' callback
+ *
+ * @param responce
+ * @param fetch
+ * @param mergeKey
+ */
+mwUtil.populate = (responce, fetch, mergeKey) => {
+    const requests = {};
+    const setters = [];
+
+    function _traverse(node, removeFromParent) {
+        function requestResource(resource) {
+            requests[resource] = requests[resource] || fetch(resource);
+            setters.push((content) => {
+                if (content[resource]) {
+                    Object.assign(node, content[resource]);
+                } else if (removeFromParent) {
+                    removeFromParent();
+                }
+                // TODO: We would normally delete it anyway,
+                // this check is here until we remove a hack with populating $merge internally
+                if (mergeKey.indexOf('$') !== -1) {
+                    delete node[mergeKey];
+                }
+            });
+        }
+
+        if (Array.isArray(node)) {
+            for (let i = 0; i < node.length; i++) {
+                _traverse(node[i], () => node.splice(i, 1));
+            }
+        } else if (typeof node === 'object') {
+            if (Array.isArray(node[mergeKey])) {
+                node[mergeKey].forEach(requestResource);
+            } else if (node[mergeKey]) {
```

Comment:
It looks like this would allow plain strings. Do we want to do that, or limit ourselves to arrays only, for simplicity and consistency?